File: app/src/main/java/dev/yiray/qickmvivm/ui/home/HomeViewState.java
package dev.yiray.qickmvivm.ui.home;

import androidx.annotation.NonNull;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import dev.yiray.qickmvivm.base.BaseState;
import dev.yiray.qickmvivm.model.Task;

public interface HomeViewState extends BaseState<HomeViewState.ViewState> {
    final class ViewState extends BaseState.BaseViewState<ViewState> implements HomeViewState {
        private String text;
        private Boolean isChecked;
        private List<Task> taskList;

        static class Builder {
            private String text;
            private Boolean isChecked;
            private List<Task> taskList;

            Builder () {
                isChecked = false;
                taskList = new ArrayList<>();
            }

            Builder(ViewState origin){
                text = origin.text;
                isChecked = origin.isChecked;
                taskList = origin.taskList;
            }

            public Builder setText(String text) {
                this.text = text;
                return this;
            }

            public Builder setChecked(boolean checked) {
                isChecked = checked;
                return this;
            }

            public Builder setTaskList(List<Task> taskList) {
                this.taskList = taskList;
                return this;
            }

            public ViewState build() {
                return new ViewState(text, isChecked, taskList);
            }
        }

        public ViewState(String text, Boolean isChecked, List<Task> taskList) {
            this.text = text;
            this.isChecked = isChecked;
            this.taskList = taskList;
        }

        public String getText() {
            return text;
        }

        public Boolean getChecked() {
            return isChecked;
        }

        public List<Task> getTaskList() {
            return taskList;
        }

        @Override
        public String toString() {
            return "ViewState{" +
                    "text='" + text + '\'' +
                    ", isChecked=" + isChecked +
                    ", taskList=" + taskList +
                    '}';
        }

        @Override
        public ViewState toViewState() {
            return this;
        }

        @Override
        public ViewState copy(ViewState nextState) {
            if (nextState == null) {
                return this;
            }

            Builder builder = new Builder(this);

            if (nextState.text != null) {
                builder.setText(nextState.text);
            }

            if (nextState.isChecked != null) {
                builder.setChecked(nextState.isChecked);
            }

            if (nextState.taskList != null) {
                builder.setTaskList(nextState.taskList);
            }

            return builder.build();
        }
    }

    final class InputState implements HomeViewState {

        private String text;

        public InputState(String text) {
            this.text = text;
        }

        @Override
        public ViewState toViewState() {
            return new ViewState.Builder().setText(text).build();
        }

        @Override
        public String toString() {
            return "InputState{" +
                    "text='" + text + '\'' +
                    '}';
        }
    }

    final class CheckBoxState implements HomeViewState {
        private boolean isChecked;

        public CheckBoxState(boolean isChecked) {
            this.isChecked = isChecked;
        }

        @Override
        public ViewState toViewState() {
            return new ViewState.Builder().setChecked(isChecked).build();
        }

        @Override
        public String toString() {
            return "CheckBoxState{" +
                    "isChecked=" + isChecked +
                    '}';
        }
    }

    final class NextPage implements HomeViewState, SideEffect {
        @Override
        public ViewState toViewState() {
            return null;
        }

        @NonNull
        @Override
        public String toString() {
            return "NextPage{}";
        }
    }

    final class ListModified implements HomeViewState {
        private int position;
        private Task task;

        public ListModified(int position, Task task) {
            this.position = position;
            this.task = task;
        }

        public Task getTask() {
            return task;
        }

        public int getPosition() {
            return position;
        }

        @Override
        public ViewState toViewState() {
            return new ViewState.Builder().setTaskList(Collections.singletonList(task)).build();
        }

        @Override
        public String toString() {
            return "ListModified{" +
                    "position=" + position +
                    ", task=" + task +
                    '}';
        }
    }
}

File: app/src/main/java/dev/yiray/livedatamvi/ui/home/HomeViewState.kt
package dev.yiray.livedatamvi.ui.home

import dev.yiray.livedatamvi.base.BaseState.SideEffect
import dev.yiray.livedatamvi.base.BaseState
import dev.yiray.livedatamvi.base.BaseState.BaseViewState
import dev.yiray.livedatamvi.model.Task

interface HomeViewState : BaseState<HomeViewState.ViewState?> {
    data class ViewState(
        val text: String? = null,
        val checked: Boolean = false,
        val taskList: List<Task?>? = null,
        val lastModifiedPosition: Int? = null
    ) :
        BaseViewState<ViewState?>(), HomeViewState {

        override fun toViewState(): ViewState {
            return this
        }

        override fun copy(nextState: ViewState?): ViewState {
            if (nextState == null) {
                return this
            }

            return copy(
                text = nextState.text ?: text,
                checked = nextState.checked,
                taskList = nextState.taskList ?: taskList,
                lastModifiedPosition = nextState.lastModifiedPosition ?: lastModifiedPosition
            )
        }
    }

    data class InputState(private val text: String) : HomeViewState {
        override fun toViewState(): ViewState {
            return ViewState(text)
        }
    }

    data class CheckBoxState(private val isChecked: Boolean) : HomeViewState {
        override fun toViewState(): ViewState {
            return ViewState(checked = isChecked)
        }
    }

    class NextPage : HomeViewState, SideEffect {
        override fun toViewState(): ViewState? {
            return null
        }

        override fun toString(): String {
            return "NextPage{}"
        }
    }

    data class ListModified(val position: Int, val task: Task?) : HomeViewState {
        override fun toViewState(): ViewState {
            return ViewState(taskList = listOf(task), lastModifiedPosition = position)
        }
    }
}